cpp/src/arrow/io/file.cc

```diff
@@ -727,6 +729,12 @@ Result<std::shared_ptr<Buffer>> MemoryMappedFile::Read(int64_t nbytes) {
   return buffer;
 }
 
+Result<Future<std::shared_ptr<Buffer>>> MemoryMappedFile::ReadAsync(int64_t position,
+                                                                    int64_t nbytes) {
+  ARROW_ASSIGN_OR_RAISE(auto buf, ReadAt(position, nbytes));
+  return Future<std::shared_ptr<Buffer>>::MakeFinished(std::move(buf));
+}
```

Comment:
> Should this API always be `Future<...>` and not `Result<Future<...>>` (even for implementations that don't actually run asynchronously)? Otherwise there are two different failure modes to take into consideration

Generally speaking, this would mean that a failure which *does* happen wholly within the scope of a `Future` returning function call would be wrapped into an error future. For example

```suggestion
Future<std::shared_ptr<Buffer>> MemoryMappedFile::ReadAsync(int64_t position,
                                                            int64_t nbytes) {
  auto maybe_buf, ReadAt(position, nbytes);
  if (maybe_buf.ok()) {
    return Future<std::shared_ptr<Buffer>>::MakeFinished(std::move(*maybe_buf));
  }
  return Future<std::shared_ptr<Buffer>>::MakeFailed(std::move(maybe_buf.status()));
}
```

This means we lose the ability to distinguish between synchronous and asynchronous errors (since `returned_future.state() == FutureState::FAILURE` might indicate either), but I think that's acceptable. @pitrou?

Reply:
Yes, I think it is. I'll make the change.